control-plane/agents/src/bin/core/volume/snapshot_operations.rs

```diff
@@ -335,88 +348,97 @@ impl OperationGuardArc<VolumeSnapshot> {
+            .collect::<Vec<_>>();
+
+        let nexus_snap_desc = replica_ids
+            .into_iter()
+            .zip(replica_snaps.clone().into_iter())
+            .map(|(x, y)| CreateNexusSnapReplDescr::new(x, y.spec().uuid().clone()))
+            .collect();
         let response = target_node
             .create_nexus_snapshot(&CreateNexusSnapshot::new(
                 SnapshotParameters::new(target.nexus(), generic_params.clone()),
-                vec![CreateNexusSnapReplDescr::new(
-                    replica_id,
-                    replica_snap.spec().uuid().clone(),
-                )],
+                nexus_snap_desc,
             ))
             .await?;
 
-        if response.skipped.contains(replica_id) || !response.skipped.is_empty() {
+        if !response.skipped.is_empty() {
             return Err(SvcError::ReplicaSnapSkipped {
-                replica: replica_snap.spec().uuid().to_string(),
+                replica: response
+                    .skipped
+                    .iter()
+                    .map(|r| r.to_string())
+                    .collect::<Vec<String>>()
+                    .join(", "),
             });
         }
-
-        let snapped = match response.replicas_status.as_slice() {
-            [snapped] if &snapped.replica_uuid == replica_snap.spec().source_id().replica_id() => {
-                Ok(snapped)
-            }
-            _ => Err(SvcError::ReplicaSnapMiss {
-                replica: replica_snap.spec().uuid().to_string(),
-            }),
-        }?;
-
-        if let Some(error) = snapped.error {
+        if let Some(snap) = response
+            .replicas_status
+            .iter()
+            .find(|&snap| snap.error.is_some())
+        {
             return Err(SvcError::ReplicaSnapError {
-                replica: replica_snap.spec().uuid().to_string(),
-                error,
+                replica: snap.replica_uuid.to_string(),
+                error: snap.error.unwrap(),
             });
         }
-
         let timestamp = DateTime::<Utc>::from(response.snap_time);
         // What if snapshot succeeds but we can't fetch the replica snapshot, should we carry
         // on as following, or should we bail out?
-        let node = registry.node_wrapper(&replica.node).await?;
-
-        let snapshot = NodeWrapper::fetch_update_snapshot_state(
-            &node,
-            ReplicaSnapshotInfo::new(
-                replica_snap.spec().source_id().replica_id(),
-                replica_snap.spec().uuid().clone(),
-            ),
-        )
-        .await?;
-
-        replica_snap.complete_vol(
-            snapshot.timestamp().into(),
-            snapshot.replica_size(),
-            snapshot.allocated_size() + snapshot.predecessor_alloc_size(),
-        );
-        Ok(VolumeSnapshotCreateResult::new_ok(replica_snap, timestamp))
+        for (replica, replica_snap) in replicas.iter().zip(replica_snaps.iter_mut()) {
```

Comment:
how do you ensure replica_snaps is associated to the equivalent replica?

Reply:
Done as vec<(state, snapshot)> now to have one to one mapping.